Compute a better `lint_node_id` during expansion

Aaron1011 opened the pull request:
When we need to emit a lint at a macro invocation, we currently use the
`NodeId` of its parent definition (e.g. the enclosing function). This
means that any `#[allow]` / `#[deny]` attributes placed 'closer' to the
macro (e.g. on an enclosing block or statement) will have no effect.

This commit computes a better `lint_node_id` in `InvocationCollector`.
When we visit/flat_map an AST node, we assign it a `NodeId` (earlier
than we normally would), and store than `NodeId` in current
`ExpansionData`. When we collect a macro invocation, the current
`lint_node_id` gets cloned along with our `ExpansionData`, allowing it
to be used if we need to emit a lint later on.

This improves the handling of `#[allow]` / `#[deny]` for
`SEMICOLON_IN_EXPRESSIONS_FROM_MACROS` and some `asm!`-related lints.
The 'legacy derive helpers' lint retains its current behavior
(I've inlined the now-removed `lint_node_id` function), since
there isn't an `ExpansionData` readily available.

rust-highfive commented:
r? @jackh726

(rust-highfive has picked a reviewer for you, use r? to override)

Aaron1011 commented:
r? @petrochenkov

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3072085256) [(plain)](https://github.com/rust-lang/rust/commit/2432650699dc465cca6abb58a0c0fd4a607e9066/checks/3072085256/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    |                          ^^^
    |
    = note: delayed at /checkout/compiler/rustc_lint/src/early.rs:394:18

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1050:13

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.55.0-nightly (0c68e3ea0 2021-07-15) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z macro-backtrace -Z tls-model=initial-exec -Z unstable-options -Z binary-dep-depinfo -Z force-unstable-if-unmarked -C opt-level=3 -C embed-bitcode=no -C debuginfo=0 -C debug-assertions=on -C overflow-checks=off -C link-args=-Wl,-rpath,$ORIGIN/../lib -C prefer-dynamic -C llvm-args=-import-instr-limit=10 --crate-type lib
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3072346532) [(plain)](https://github.com/rust-lang/rust/commit/3d3f11e91315224d91b0485b9480238e31370ba1/checks/3072346532/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_lint/src/early.rs at line 124:
         self.with_lint_attrs(f.id, &f.attrs, |cx| {
             ast_visit::walk_expr_field(cx, f);
-
     }
 
 
     fn visit_stmt(&mut self, s: &'a ast::Stmt) {
Diff in /checkout/compiler/rustc_lint/src/early.rs at line 398:
     // that was not lint-checked (perhaps it doesn't exist?). This is a bug.
     for (id, lints) in buffered.map {
         for early_lint in lints {
-            sess.delay_span_bug(early_lint.span, &format!("failed to process buffered lint here (dummy = {})", id == ast::DUMMY_NODE_ID));
+            sess.delay_span_bug(
+                early_lint.span,
+                &format!(
+                    "failed to process buffered lint here (dummy = {})",
+                    id == ast::DUMMY_NODE_ID
+            );
         }
     }
 }
 }
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_ast/src/util/comments/tests.rs" "/checkout/compiler/rustc_ast/src/attr/mod.rs" "/checkout/compiler/rustc_ast/src/ast.rs" "/checkout/compiler/rustc_lint/src/non_ascii_idents.rs" "/checkout/compiler/rustc_lint/src/traits.rs" "/checkout/compiler/rustc_lint/src/methods.rs" "/checkout/compiler/rustc_lint/src/early.rs" "/checkout/compiler/rustc_lint/src/lib.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3072842873) [(plain)](https://github.com/rust-lang/rust/commit/1f2711b67e7889be10f376cc69f08706009e379f/checks/3072842873/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Check compiletest suite=run-make-fulldeps mode=run-make (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 224 tests
..........i.ii...ii................................................................................. 100/224
...............i..................iiiiii...F..i.................iii...........ii.................... 200/224
Some tests failed in compiletest suite=run-make-fulldeps mode=run-make host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
failures:

---- [run-make] run-make-fulldeps/pretty-expanded stdout ----
---- [run-make] run-make-fulldeps/pretty-expanded stdout ----

error: make failed
status: exit status: 2
command: "make"
stdout:
------------------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pretty-expanded/pretty-expanded:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pretty-expanded/pretty-expanded -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pretty-expanded/pretty-expanded  -o /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pretty-expanded/pretty-expanded/input.expanded.rs -Z unstable-options \
 --pretty=expanded input.rs
Makefile:4: recipe for target 'all' failed
------------------------------------------
stderr:
------------------------------------------
warning: ignoring --out-dir flag due to -o flag
warning: ignoring --out-dir flag due to -o flag

warning: variant `y` should have an upper camel case name
  --> input.rs:11:40
   |
11 | #[derive(RustcEncodable)] pub enum E { y }
   |                                        ^ help: convert the identifier to upper camel case (notice the capitalization): `Y`
   = note: `#[warn(non_camel_case_types)]` on by default


warning: variant `z` should have an upper camel case name
  --> input.rs:12:40
   |
12 | #[derive(RustcEncodable)] pub enum F { z(isize) }
   |                                        ^ help: convert the identifier to upper camel case (notice the capitalization): `Z`
warning: 3 warnings emitted


error: internal compiler error: failed to process buffered lint here (dummy = false)
 --> input.rs:7:10
  |
7 | #[derive(RustcEncodable)] pub struct A;
  |
  = note: delayed at /checkout/compiler/rustc_lint/src/early.rs:400:18


error: internal compiler error: failed to process buffered lint here (dummy = false)
 --> input.rs:8:10
  |
8 | #[derive(RustcEncodable)] pub struct B(isize);
  |
  = note: delayed at /checkout/compiler/rustc_lint/src/early.rs:400:18


error: internal compiler error: failed to process buffered lint here (dummy = false)
 --> input.rs:9:10
  |
9 | #[derive(RustcEncodable)] pub struct C { x: isize }
  |
  = note: delayed at /checkout/compiler/rustc_lint/src/early.rs:400:18


error: internal compiler error: failed to process buffered lint here (dummy = false)
  --> input.rs:10:10
   |
10 | #[derive(RustcEncodable)] pub enum D {}
   |
   = note: delayed at /checkout/compiler/rustc_lint/src/early.rs:400:18


error: internal compiler error: failed to process buffered lint here (dummy = false)
  --> input.rs:11:10
   |
11 | #[derive(RustcEncodable)] pub enum E { y }
   |
   = note: delayed at /checkout/compiler/rustc_lint/src/early.rs:400:18


error: internal compiler error: failed to process buffered lint here (dummy = false)
  --> input.rs:12:10
   |
12 | #[derive(RustcEncodable)] pub enum F { z(isize) }
   |
   = note: delayed at /checkout/compiler/rustc_lint/src/early.rs:400:18


thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1050:13

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.
---
note: compiler flags: -Z unstable-options

query stack during panic:
end of query stack
make: *** [all] Error 101
------------------------------------------




failures:
    [run-make] run-make-fulldeps/pretty-expanded

test result: FAILED. 203 passed; 1 failed; 20 ignored; 0 measured; 0 filtered out; finished in 23.21s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--rust-demangler-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/rust-demangler" "--src-base" "/checkout/src/test/run-make-fulldeps" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "run-make-fulldeps" "--mode" "run-make" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--llvm-bin-dir" "/usr/lib/llvm-10/bin" "--cc" "cc" "--cxx" "c++" "--cflags" "-ffunction-sections -fdata-sections -fPIC -m64" "--ar" "ar" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:27:02
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3079554359) [(plain)](https://github.com/rust-lang/rust/commit/6b9b1ae9756abe8dacdcbdb2d303a00bdf4ee353/checks/3079554359/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Check compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 12142 tests
.................................................................................................... 100/12142
.............................................ii.............iiF..................................... 200/12142
.................................................................................................... 400/12142
.................................................................................................... 500/12142
.................................................................................................... 600/12142
...............................................i.................................................... 700/12142
---
.............ii.ii.......i...i...................................................................... 6500/12142
............................................................i....i.................................. 6600/12142
..i...........................i..................................................................... 6700/12142
..........................................................................i......................... 6800/12142
............F..F.................................................................................... 6900/12142
................................i................................................................... 7100/12142
.................................................................................................... 7200/12142
.................................................................................................... 7300/12142
.................................................................................................... 7400/12142
---
.................................i.i................................................................ 12100/12142
..........................................
failures:

---- [ui] ui/asm/inline-syntax.rs#x86_64 stdout ----


1 warning: avoid using `.intel_syntax`, Intel syntax is the default
+    |
+    |
+ LL | global_asm!(".intel_syntax noprefix", "nop");
+    |
+    = note: `#[warn(bad_asm_style)]` on by default
+ 
+ 
+ warning: avoid using `.intel_syntax`, Intel syntax is the default
3    |
3    |
4 LL |         asm!(".intel_syntax noprefix", "nop");
5    |               ^^^^^^^^^^^^^^^^^^^^^^
-    |
-    = note: `#[warn(bad_asm_style)]` on by default
8 
8 
9 warning: avoid using `.intel_syntax`, Intel syntax is the default

35    |
35    |
36 LL |             .intel_syntax noprefix
- 
- 
- warning: avoid using `.intel_syntax`, Intel syntax is the default
-   --> $DIR/inline-syntax.rs:57:14
-    |
- LL | global_asm!(".intel_syntax noprefix", "nop");
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
44 
45 warning: 7 warnings emitted
46 
46 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/inline-syntax.x86_64/inline-syntax.x86_64.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args asm/inline-syntax.rs`

error in revision `x86_64`: 1 errors occurred comparing output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/asm/inline-syntax.rs" "-Zthreads=1" "--cfg" "x86_64" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/inline-syntax.x86_64" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target" "x86_64-unknown-linux-gnu" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/inline-syntax.x86_64/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
warning: avoid using `.intel_syntax`, Intel syntax is the default
   |
   |
LL | global_asm!(".intel_syntax noprefix", "nop");
   |
   = note: `#[warn(bad_asm_style)]` on by default


warning: avoid using `.intel_syntax`, Intel syntax is the default
   |
   |
LL |         asm!(".intel_syntax noprefix", "nop");


warning: avoid using `.intel_syntax`, Intel syntax is the default
   |
   |
LL |         asm!(".intel_syntax aaa noprefix", "nop");


warning: avoid using `.att_syntax`, prefer using `options(att_syntax)` instead
   |
   |
LL |         asm!(".att_syntax noprefix", "nop");


warning: avoid using `.att_syntax`, prefer using `options(att_syntax)` instead
   |
   |
LL |         asm!(".att_syntax bbb noprefix", "nop");


warning: avoid using `.intel_syntax`, Intel syntax is the default
   |
   |
LL |         asm!(".intel_syntax noprefix; nop");


warning: avoid using `.intel_syntax`, Intel syntax is the default
   |
   |
LL |             .intel_syntax noprefix

warning: 7 warnings emitted



------------------------------------------


---- [ui] ui/lint/semicolon-in-expressions-from-macros/allow-semicolon-in-expressions-from-macros.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/lint/semicolon-in-expressions-from-macros/allow-semicolon-in-expressions-from-macros.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/semicolon-in-expressions-from-macros/allow-semicolon-in-expressions-from-macros" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/semicolon-in-expressions-from-macros/allow-semicolon-in-expressions-from-macros/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: internal compiler error: failed to process buffered lint here (dummy = false)
   |
LL |         true;
   |             ^
...
...
LL |         _ => foo!()
   |              ------ in this macro invocation
   |
   = note: delayed at /checkout/compiler/rustc_lint/src/early.rs:400:18
   = note: this error: internal compiler error originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1050:13

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.55.0-nightly (c1a35a796 2021-07-15) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z emit-future-incompat-report -C codegen-units=1 -C prefer-dynamic -C rpath -C debuginfo=0
query stack during panic:
end of query stack

------------------------------------------
---

4 LL |         true;
5    |             ^
6 ...
- LL |         foo!(warn_in_block)
-    |         ------------------- in this macro invocation
+ LL |     let _ = foo!(second);
9    |
10 note: the lint level is defined here
11   --> $DIR/semicolon-in-expressions-from-macros.rs:4:9

---
+    |
+ LL |         true;
+    |             ^
+ ...
+ LL |         foo!(warn_in_block)
+    |
+    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
+    = note: for more information, see issue #79813 <https://github.com/rust-lang/rust/issues/79813>
+    = note: this warning originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)
---
To only update this specific test, also pass `--test-args lint/semicolon-in-expressions-from-macros/semicolon-in-expressions-from-macros.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/lint/semicolon-in-expressions-from-macros/semicolon-in-expressions-from-macros.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/semicolon-in-expressions-from-macros/semicolon-in-expressions-from-macros" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2018" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/semicolon-in-expressions-from-macros/semicolon-in-expressions-from-macros/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
warning: trailing semicolon in macro used in expression position
  --> /checkout/src/test/ui/lint/semicolon-in-expressions-from-macros/semicolon-in-expressions-from-macros.rs:9:13
   |
LL |         true; //~  WARN trailing semicolon in macro
...
...
LL |     let _ = foo!(second);
   |
note: the lint level is defined here
  --> /checkout/src/test/ui/lint/semicolon-in-expressions-from-macros/semicolon-in-expressions-from-macros.rs:4:9
   |
---

warning: trailing semicolon in macro used in expression position
  --> /checkout/src/test/ui/lint/semicolon-in-expressions-from-macros/semicolon-in-expressions-from-macros.rs:9:13
   |
LL |         true; //~  WARN trailing semicolon in macro
...
...
LL |     let _ = foo!(warn_in_expr);
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #79813 <https://github.com/rust-lang/rust/issues/79813>
   = note: this warning originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)
   = note: this warning originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: trailing semicolon in macro used in expression position
  --> /checkout/src/test/ui/lint/semicolon-in-expressions-from-macros/semicolon-in-expressions-from-macros.rs:9:13
   |
LL |         true; //~  WARN trailing semicolon in macro
...
...
LL |     let _ = #[allow(semicolon_in_expressions_from_macros)] foo!(allow_does_not_work);
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #79813 <https://github.com/rust-lang/rust/issues/79813>
   = note: this warning originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)
   = note: this warning originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: trailing semicolon in macro used in expression position
  --> /checkout/src/test/ui/lint/semicolon-in-expressions-from-macros/semicolon-in-expressions-from-macros.rs:9:13
   |
LL |         true; //~  WARN trailing semicolon in macro
...
...
LL |         foo!(warn_in_block)
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #79813 <https://github.com/rust-lang/rust/issues/79813>
   = note: this warning originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)
---
test result: FAILED. 12038 passed; 3 failed; 101 ignored; 0 measured; 0 filtered out; finished in 120.63s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:13:29
```

</details>

Aaron1011 commented:
@petrochenkov: This PR is now ready for review

petrochenkov commented:
There are about 30 `NodeId`s in AST, but `assign_id!` is used 19 times in this PR.
This means only *some* `NodeId`s can be assigned to `lint_node_id`.
I'm not sure which subset of `NodeId`s is used for linting in this PR, and which subset *should* be used.

I think it's enough to track IDs that can be *immediate parents* of macro invocations (with "macro invocations" including uses of macro attributes).

If we are considering attributes, then some interesting situations may arise.
For example, does this example produce a warning?
```rustc
#[allow(deprecated)]
#[deprecated_macro_attribute_equivalent_to_cfg_false]
struct S;
```
Apparently it does, because `struct S;` disappears before getting a meaningful `NodeId` (the ID is assigned after expanding all macro attributes), so `struct S;` is not an immediate parent of the macro attribute invocation, the containing module/crate is.
(Anyway, this snippet probably needs to be added as a test case.)

petrochenkov commented:
>I think it's enough to track IDs that can be immediate parents of macro invocations

Looks like I was wrong, pretty much anything with a `NodeId` can be an immediate parent of a macro invocation (including stuff like path segments - `seg<ty!()>`, an excluding only the most basic nodes like lifetimes).

However, it's not necessary to track them all, we only need to track parent nodes (not necessarily immediate) supporting attributes like `#[allow]`, i.e. only nodes listed in `Annotatable` (that's only 13 `assign_id!`s), tracking with higher granularity won't have any effect.

petrochenkov commented on compiler/rustc_expand/src/expand.rs:
```diff
@@ -1118,12 +1153,19 @@ impl<'a, 'b> MutVisitor for InvocationCollector<'a, 'b> {
+    // This is needed in order to set `lint_node_id` for `let` statements
+    fn visit_local(&mut self, local: &mut P<Local>) {
+        assign_id!(self, &mut local.id, || noop_visit_local(local, self));
+    }
```
Looks like this shouldn't be necessary, tracking the `Stmt`'s ID should be enough (https://github.com/rust-lang/rust/pull/87146#issuecomment-881890210), the buffered lint reporter in `compiler/rustc_lint/src/early.rs` does seem to associate lints with `Stmt` IDs.

petrochenkov commented on compiler/rustc_expand/src/expand.rs:
```diff
@@ -1225,7 +1269,9 @@ impl<'a, 'b> MutVisitor for InvocationCollector<'a, 'b> {
-            _ => return noop_visit_pat(pat, self),
+            _ => {
+                return assign_id!(self, &mut pat.id, || noop_visit_pat(pat, self));
+            }
```
This shouldn't be necessary because patterns do not support attributes (https://github.com/rust-lang/rust/pull/87146#issuecomment-881890210).

petrochenkov commented on compiler/rustc_expand/src/expand.rs:
```diff
@@ -1278,7 +1326,7 @@ impl<'a, 'b> MutVisitor for InvocationCollector<'a, 'b> {
             DirOwnership::UnownedViaBlock,
         );
-        noop_visit_block(block, self);
+        assign_id!(self, &mut block.id, || noop_visit_block(block, self));
```
This shouldn't be necessary because blocks by themselves do not support attributes (https://github.com/rust-lang/rust/pull/87146#issuecomment-881890210).

petrochenkov commented on compiler/rustc_expand/src/expand.rs:
```diff
@@ -1434,14 +1489,16 @@ impl<'a, 'b> MutVisitor for InvocationCollector<'a, 'b> {
         match ty.kind {
             ast::TyKind::MacCall(_) => {}
-            _ => return noop_visit_ty(ty, self),
+            _ => return assign_id!(self, &mut ty.id, || noop_visit_ty(ty, self)),
```
This shouldn't be necessary because types do not support attributes (https://github.com/rust-lang/rust/pull/87146#issuecomment-881890210).

petrochenkov commented on compiler/rustc_expand/src/expand.rs:
```diff
@@ -1266,6 +1312,8 @@ impl<'a, 'b> MutVisitor for InvocationCollector<'a, 'b> {
 
         // The placeholder expander gives ids to statements, so we avoid folding the id here.
+        // We don't use `assign_id!` - it will be called when we visit statement's contents
+        // (e.g. an expression, item, or local)
```
It may be more consistent to use `assign_id!` with a `Stmt` instead, because `Stmt` is an attribute target (https://github.com/rust-lang/rust/pull/87146#issuecomment-881890210).

Aaron1011 replied:
For `StmtKind::Item` and `StmtKind::Expr`, we will immediately overwrite the tracked statement ID with the item/expr id. Also, `Stmt` id assignment is performed in `PlaceholderExpander` instead of `InvocationCollector`, so using the current approach allows us to keep that logic as-is.

Aaron1011 commented:
@petrochenkov: I've added the `#[allow(deprecated)]` test case, and removed the unnecessary `assign_id!` calls. See my reply to your comment about the handling of `Stmt`

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit cb862d5cac071a21735886dac9878ffb4562ae9f has been approved by `petrochenkov`

<!-- @bors r=petrochenkov cb862d5cac071a21735886dac9878ffb4562ae9f -->
<!-- homu: {"type":"Approved","sha":"cb862d5cac071a21735886dac9878ffb4562ae9f","approver":"petrochenkov"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #86676) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Aaron1011 commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit 1c1c7949ab6a8d4150b7e645ae4779677e86ebd1 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 1c1c7949ab6a8d4150b7e645ae4779677e86ebd1 -->
<!-- homu: {"type":"Approved","sha":"1c1c7949ab6a8d4150b7e645ae4779677e86ebd1","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 1c1c7949ab6a8d4150b7e645ae4779677e86ebd1 with merge 9cc5500d72f84850fbde80fdf5520e2c97c1626d...
<!-- homu: {"type":"BuildStarted","head_sha":"1c1c7949ab6a8d4150b7e645ae4779677e86ebd1","merge_sha":"9cc5500d72f84850fbde80fdf5520e2c97c1626d"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

Aaron1011 commented:
@bors retry

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3098804385) [(plain)](https://github.com/rust-lang-ci/rust/commit/9cc5500d72f84850fbde80fdf5520e2c97c1626d/checks/3098804385/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:hourglass: Testing commit 1c1c7949ab6a8d4150b7e645ae4779677e86ebd1 with merge 0ecff8c623fa29cccad6f946c657e81dacd85847...
<!-- homu: {"type":"BuildStarted","head_sha":"1c1c7949ab6a8d4150b7e645ae4779677e86ebd1","merge_sha":"0ecff8c623fa29cccad6f946c657e81dacd85847"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3101418690)
Approved by: petrochenkov
Pushing 0ecff8c623fa29cccad6f946c657e81dacd85847 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3101418690"},"merge_sha":"0ecff8c623fa29cccad6f946c657e81dacd85847"} -->
